File: shelf/src/main/java/com/toddway/shelf/serializer/GsonSerializer.java
package com.toddway.shelf.serializer;

import com.google.gson.Gson;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.OutputStreamWriter;

/**
 * Created by tway on 4/7/16.
 */
public class GsonSerializer implements Serializer {

    Gson gson;

    public GsonSerializer() {
        gson = new Gson();
    }

    public GsonSerializer(Gson gson) {
        this.gson = gson;
    }

    @Override
    public <T> T deserialize(InputStream inputStream, Class<T> type) {
        return gson.fromJson(new JsonReader(new InputStreamReader(inputStream)), type);
    }

    @Override
    public <T> void serialize(OutputStream outputStream, T object) {
        JsonWriter writer = new JsonWriter(new OutputStreamWriter(outputStream));
        gson.toJson(object, object.getClass(), writer);
        try {
            writer.close();
        } catch (IOException e) {

        }
    }
}

File: shelf/src/jvmMain/kotlin/com/toddway/shelf/GsonSerializer.kt
package com.toddway.shelf

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import kotlin.reflect.KClass

class GsonSerializer(val gson : Gson = Gson()) : Shelf.Serializer {
    override fun <T : Any> fromType(value: T): String {
        return gson.toJson(value)
    }

    override fun <T : Any> toType(string: String, klass: KClass<T>): T {
        return gson.fromJson(string, klass.java)
    }

    override fun <T : Any> toTypeList(string: String, klass: KClass<T>): List<T> {
        return gson.fromJson<List<T>>(string, listType(klass.java))
    }

    private fun <T : Any> listType(type : Class<T>) =
        TypeToken.getParameterized(List::class.java, type).type
}